net/tsdial/tsdial.go

```diff
@@ -139,15 +139,39 @@ func (d *Dialer) SetNetMon(netMon *netmon.Monitor) {
 }
 
 func (d *Dialer) linkChanged(delta *netmon.ChangeDelta) {
-	if !delta.Major {
-		return
-	}
 	d.mu.Lock()
 	defer d.mu.Unlock()
 	for id, c := range d.activeSysConns {
-		go c.Close()
-		delete(d.activeSysConns, id)
+		if changeAffectsConn(delta, c) {
+			go c.Close()
+			delete(d.activeSysConns, id)
+		}
+	}
+}
+
+// changeAffectsConn reports whether the network change delta affects
+// the provided connection.
+func changeAffectsConn(delta *netmon.ChangeDelta, conn net.Conn) bool {
+	la, _ := conn.LocalAddr().(*net.TCPAddr)
+	ra, _ := conn.RemoteAddr().(*net.TCPAddr)
+	if la == nil || ra == nil {
+		return false // not TCP
+	}
+	lip, rip := la.AddrPort().Addr(), ra.AddrPort().Addr()
+
+	if delta.Old == nil {
+		return false
+	}
+	if delta.Old.DefaultRouteInterface != delta.New.DefaultRouteInterface ||
+		delta.Old.HTTPProxy != delta.New.HTTPProxy {
+		return true
+	}
+	if !delta.New.HasIP(lip) && delta.Old.HasIP(lip) {
+		// Our interface with this source IP went away.
+		return true
 	}
+	_ = rip // TODO(bradfitz): use the remote IP?
+	return false
```

Comment:
Could check if the remote addr uses a non-default interface, like if a longer prefix matches or something. That may already be handled by the local ip check though.